### PR TITLE
Cheddar token review + user feedback

### PR DESCRIPTION
**Pull request by robert-zaremba**
+ Fixed function scope modifiers. Many internal functions were exposed as public, but they should be private.
+ removed `default_ft_metadata` and moved the default metadata as an initial metadata value in the constructor.
+ added logs and comments to `int_ft_resolve_transfer`  and renamed it to `ft_resolve_transfer_adjust`
+ removed `env_log` macro and use `log` macro from the SDK instead. 
+ added `total_staked` view method to `p1` staking pool

BREAKING:
+ changed return values of `p1.status`  function - added `total_stake` as a return value.

### CODE REVIEW

File: cheddar/src/internal.rs [hunk 1]
```diff
@@ -3,59 +3,44 @@ use near_sdk::{AccountId, Balance, PromiseResult};
...
-        symbol: "CHDR".to_string(),
-        icon: Some(String::from(
-            r###"<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 56 56"><style>.a{fill:#F4C647;}.b{fill:#EEAF4B;}</style><path d="M45 19.5v5.5l4.8 0.6 0-11.4c-0.1-3.2-11.2-6.7-24.9-6.7 -13.7 0-24.8 3.6-24.9 6.7L0 32.5c0 3.2 10.7 7.1 24.5 7.1 0.2 0 0.3 0 0.5 0V21.5l-4.7-7.2L45 19.5z" class="a"/><path d="M25 31.5v-10l-4.7-7.2L45 19.5v5.5l-14-1.5v10C31 33.5 25 31.5 25 31.5z" fill="#F9E295"/><path d="M24.9 7.5C11.1 7.5 0 11.1 0 14.3s10.7 7.2 24.5 7.2c0.2 0 0.3 0 0.5 0l-4.7-7.2 25 5.2c2.8-0.9 4.4-4 4.4-5.2C49.8 11.1 38.6 7.5 24.9 7.5z" class="b"/><path d="M36 29v19.6c8.3 0 15.6-1 20-2.5V26.5L31 23.2 36 29z" class="a"/><path d="M31 23.2l5 5.8c8.2 0 15.6-1 19.9-2.5L31 23.2z" class="b"/><polygon points="36 29 36 48.5 31 42.5 31 23.2 " fill="#FCDF76"/></svg>"###,
-        )),
-        reference: None, // TODO
-        reference_hash: None,
-        decimals: 24,
-    }
-}
-
```
Comment (robert-zaremba) [inline cheddar/src/internal.rs:19]: we don't need this function, we can use it as an initial value in the constructor and optimize the `internal_get_ft_metadata`.

File: cheddar/src/internal.rs [hunk 2]
```diff
@@ -139,27 +130,36 @@ impl Contract {
...
                     // Sender's account was deleted, so we need to burn tokens.
-                    self.total_supply -= refund_amount;
-                    log!("The account of the sender was deleted");
-                    return (amount, refund_amount);
+                    self.total_supply -= unused_amount;
+                    log!(
+                        "The sender account is deleted, can't make a refund. Burning {} from ft_transfer_call",
+                        unused_amount
+                    );
+                    return (amount - unused_amount, unused_amount);
```
Comment (robert-zaremba) [inline cheddar/src/internal.rs:158]: FIX: the return value was wrong here.

File: cheddar/src/internal.rs (hunk 1 above)
Comment (luciotato) [reply cheddar/src/internal.rs:19]: The you'll have it in the code *and* in the storage. The reason why it's there is to avoid occupying storage with something you can generate.
Said that, I've no problem if you want to change this. Both mechanisms have pros and cons.

File: p1-staking-pool-dyn/src/lib.rs [hunk 3]
```diff
@@ -95,16 +96,23 @@ impl Contract {
...
     }
 
-    /// Returns amount of staked NEAR, farmed CHEDDAR and the current round.
-    pub fn status(&self, account_id: AccountId) -> (U128, U128, u64) {
+    /// Returns amount of total staked NEAR, user staked NEAR, farmed CHEDDAR
+    /// and the current round.
+    pub fn status(&self, account_id: AccountId) -> (U128, U128, U128, u64) {
         let mut v = self.get_vault_or_default(&account_id);
         return (
+            self.total_stake.into(),
```
Comment (luciotato) [inline p1-staking-pool-dyn/src/lib.rs]: This breaks the UI. 
This does not go here.
"fn status" is the account status, should not include contract information.
please undo this change because it breaks the UI
`pub fn total_staked(&self) -> U128` is ok, use that

File: cheddar/src/internal.rs [hunk 4]
```diff
@@ -105,27 +90,33 @@ impl Contract {
...
         receiver_id: ValidAccountId,
         amount: U128,
     ) -> (u128, u128) {
-        let sender_id: AccountId = sender_id.into();
         let receiver_id: AccountId = receiver_id.into();
         let amount: Balance = amount.into();
 
         // Get the unused amount from the `ft_on_transfer` call result.
-        let unused_amount = match env::promise_result(0) {
+        let mut unused_amount = match env::promise_result(0) {
```
Comment (luciotato) [inline cheddar/src/internal.rs:119]: are you changing something?
it's not good to use a `mut` var when you can avoid it

File: cheddar/src/internal.rs [hunk 5]
```diff
@@ -139,27 +130,36 @@ impl Contract {
...
                         receiver_id,
                         sender_id
                     );
-                    return (amount - refund_amount, 0);
+                    return (amount - unused_amount, 0);
                 } else {
                     // Sender's account was deleted, so we need to burn tokens.
-                    self.total_supply -= refund_amount;
-                    log!("The account of the sender was deleted");
-                    return (amount, refund_amount);
```
Comment (luciotato) [inline cheddar/src/internal.rs:161]: the solution was to put `amount - refund_amount` here.
Not change the entire fn

File: cheddar/src/internal.rs (hunk 1 above)
Comment (robert-zaremba) [reply cheddar/src/internal.rs:19]: we pay for the contract storage as well, this increases the contract storage. In general it's very small, and the calls for metadata will be more expensive.

File: p1-staking-pool-dyn/src/lib.rs (hunk 3 above)
Comment (robert-zaremba) [reply p1-staking-pool-dyn/src/lib.rs]: yes, I made a note about it in the PR description. The change is motivated by the UI update. I wanted to reduce amount of calls to the network. Let's discuss on in chat to make a decision if we want this change or not.

File: cheddar/src/internal.rs (hunk 4 above)
Comment (robert-zaremba) [reply cheddar/src/internal.rs:119]: yes, later in the code there is `refund_amount` which is basically "fixed" unused amount. It was easier to reason that we are adjusting the `unused_amount` based on the recipient balance and removing one more variable to remember.

File: cheddar/src/internal.rs (hunk 5 above)
Comment (robert-zaremba) [reply cheddar/src/internal.rs:161]: the change in this function is:
+ reuse `unused_amount` -- as we fix it so no need to confuse the reader with a new variable
+ fix the return value
+ added logs to explain to the user that we were not able to correctly make a refund.

Comment (luciotato) [reply cheddar/src/internal.rs:161]: no problem, let's do it this way, but it's easier for the reader if you use immutable variables, that's the reason why that's the default. It's way easier to read and reason about the code if vars are immutable.
